### PR TITLE
feat(build): introduce Typescript and Storybook to the build process

### PR DESCRIPTION
**Pull request by austin94**
## Description

This PR focuses on adding two new technologies into our stack, [Typescript](https://www.typescriptlang.org/) and [Storybook](https://storybook.js.org/).

### Storybook

We've been talking about this for a while, but with the addition of Typescript, styleguidist is making less sense for our "visual unit test" documentation. Storybook will be able to act as a decent stop-gap until we are able to invest more time into a custom documentation solution.

This storybook config includes:

- `addon-notes` Allows us to include markdown notes with each story
- `addon-knobs` Allows us to include customizations for each story. These can be shared with a custom URL as well.
- `addon-storysource` Show the source of each story
- `addon-actions` Show actions performed within a story
- `addon-a11y` Show axe audits and color-blindness filters in preview
- `addon-contexts` Show custom `RTL` and `LTR` locale dropdown for our `ThemeProvider`

In the webpack config I have enabled the following:

- Our standard babel config
- All TS files are passed through the `babel-loader` to ensure IE11 is supported
- All files are passed through `eslint-loader` and `ts-loader` to show all type and lint errors during the storybook build 

storybookExample

storybook-example

#### Overall thoughts

Being able to write the examples in TS/JS rather than MD is a _much better_ experience. The prop tables and the Docs functionality that is in technical preview may help in the future as well.

### Typescript

The main goal of this addition is to include the `typings/index.d.ts` type declarations for the dropdowns package. This is automatically consumable for TS consumers and will benefit our standard users with an improvement in IDEs and tools that leverage the TS language server (mostly VS Code). 

The changes to introduce this were fairly minimal and included the following:

- Modify `eslint` to support linting TS file
- Prettier already handles TS well
- Add `ts-jest` to allow our jest environment to consume and type-check our tests
- Adding in all of our missing types for storybook, react, prop-types, etc.
- Adding `@babel/preset-typescript` to our standard babel config to allow TS in our standard build steps
- Add TS linting and testing steps to travis

## Try it at home

If you want to try these changes locally I have included some new npm scripts.

- `yarn start:storybook`
  - Starts storybook in development mode
- `yarn build:storybook`
  - Creates a static build of storybook. Added with our standard `build:demo` steps now
- `yarn ts`
  - Type-check all files
- `yarn ts:watch`
  - Type-check all files in watch mode (honestly this isn't needed often due to storybook and jest running typescript in their loaders)

**Example TS failure**

exampleTsFailure

If VS Code isn't showing valid syntax highlighting you may need to restart the TS language service with:

```bash
cmd + shift + p -> search for "Restart TS Server"
```

## Checklist

- [ ] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [x] :nail_care: view component styling is based on a Garden CSS
      component
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

**Comment by austin94** [comment]
I've broken this PR into 2 commits:
- Adding TS and Storybook
- The larger bulk of changes, migrating Dropdowns to TS

A great example of the benefits of TS is the ability to _actually_ refactor variables and enums. Here is a dummy example:

**Comment by sunesimonsen** [comment]
This is a pretty insane PR :-) I'm curious, can't storybook work without TypeScript? It seems like two pretty big changes to mix together.

**Comment by sunesimonsen** [comment]
I don't think you should try to split the PR, but for another time, it would be great to avoid mixing things this big into one PR :-)

Exciting times 🎉

File: .storybook/general-stories.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import React from 'react';
+import { storiesOf } from '@storybook/react';
+import { StyledMarkdown } from 'storybook-utils';
+
+import README from '../README.md';
+import CHANGELOG from '../CHANGELOG.md';
+import API from '../API.md';
+
+storiesOf('General', module)
+  .add('README', () => <StyledMarkdown dangerouslySetInnerHTML={{ __html: README }} />, {
```
Comment (ryanseddon) [inline .storybook/general-stories.tsx]: Is this just a temporary thing until storybook docs and mdx are stable?

File: .storybook/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+const path = require('path');
+const webpack = require('webpack');
+const { readdir } = require('fs').promises;
```
Comment (ryanseddon) [inline .storybook/webpack.config.js]: I had node v8 as my default and this threw an error, `lts/*` is supposed to be 10 but nvm didn't pickup the difference when i `cd`'ed into your branch. It also warns in node 10 that this is an experimental feature should we rely on this?

File: .travis.yml [hunk 3]
```diff
@@ -16,7 +16,9 @@ before_install:
 install: yarn --frozen-lockfile --ignore-scripts
 
 script:
+  - yarn lerna bootstrap --ignore-scripts
```
Comment (ryanseddon) [inline .travis.yml]: This is a duplicate of line 23

File: .vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+  "typescript.tsdk": "node_modules/typescript/lib"
```
Comment (ryanseddon) [inline .vscode/settings.json]: I don't think we should be committing editor specific stuff

File: .yarnclean [hunk 5]
```diff
@@ -0,0 +1 @@
+@types/react-native
```
Comment (ryanseddon) [inline .yarnclean]: Should this be here?

File: packages/dropdowns/src/Autocomplete/Autocomplete.spec.tsx [hunk 6]
```diff
@@ -7,7 +7,7 @@
 
 import React from 'react';
 import { render, fireEvent } from 'garden-test-utils';
-import { Dropdown, Autocomplete, Field, Menu, Item, Label } from '../';
+import { Dropdown, Autocomplete, Field, Menu, Item, Label } from '..';
```
Comment (ryanseddon) [inline packages/dropdowns/src/Autocomplete/Autocomplete.spec.tsx]: Didn't realise you could drop the `/`, neat.

File: packages/dropdowns/src/Autocomplete/Autocomplete.spec.tsx [hunk 7]
```diff
@@ -34,7 +34,7 @@ describe('Autocomplete', () => {
 
     fireEvent.click(getByTestId('autocomplete'));
 
-    expect(document.activeElement.nodeName).toBe('INPUT');
+    expect(document.activeElement!.nodeName).toBe('INPUT');
```
Comment (ryanseddon) [inline packages/dropdowns/src/Autocomplete/Autocomplete.spec.tsx]: Do these really add anything? The tests will fail if activeElement is null or undefined anyway?

File: packages/dropdowns/src/Autocomplete/Autocomplete.tsx [hunk 8]
```diff
@@ -48,7 +60,7 @@ const Autocomplete = ({ children, ...props }) => {
   const selectProps = getToggleButtonProps({
     onKeyDown: e => {
       if (isOpen) {
-        e.nativeEvent.preventDownshiftDefault = true;
+        (e.nativeEvent as any).preventDownshiftDefault = true;
```
Comment (ryanseddon) [inline packages/dropdowns/src/Autocomplete/Autocomplete.tsx]: To me non-null operator would make more sense here than the extra parentheses.

```
e.nativeEvent!.preventDownshiftDefault = true;
```

That's the same as above right?

File: packages/dropdowns/src/Dropdown/Dropdown.tsx [hunk 9]
```diff
@@ -7,18 +7,42 @@
...
+  selectedItems?: any[];
+  highlightedIndex?: number;
+  inputValue?: string;
+  onSelect?: (selectedItem: any | null, stateAndHelpers: ControllerStateAndHelpers<any>) => void;
+  onStateChange?: (
+    options: StateChangeOptions<any>,
+    stateAndHelpers: ControllerStateAndHelpers<any>
+  ) => void;
+  downshiftProps?: object;
+}
```
Comment (ryanseddon) [inline packages/dropdowns/src/Dropdown/Dropdown.tsx]: Should we have a bit more organisation around where interfaces should live? Should it be in the file itself or live along side it in another file?

File: packages/dropdowns/src/Autocomplete/Autocomplete.spec.tsx (hunk 7 above)
Comment (ryanseddon) [reply packages/dropdowns/src/Autocomplete/Autocomplete.spec.tsx]: I see, my vim typescript setup warns about possibly null value. Neat.

File: packages/dropdowns/src/stories/props.md [hunk 10]
```diff
@@ -0,0 +1,224 @@
+## `@zendeskgarden/react-dropdowns` Components
+
+- [Dropdown](#dropdown)
```
Comment (ryanseddon) [inline packages/dropdowns/src/stories/props.md]: Yikes is this automagically generated?

File: .storybook/webpack.config.js (hunk 2 above)
Comment (sunesimonsen) [reply .storybook/webpack.config.js]: You can just use:

```
const { promisify } = require('util')
const fs = require('fs')
const readdirAsync = promisify(fs.readdir)
```

File: package.json [hunk 11]
```diff
@@ -5,36 +5,67 @@
...
     "lint": "yarn lint:css && yarn lint:js && yarn lint:markdown",
     "lint:css": "stylelint \"packages/*/src/**/*.js\" && stylelint \"utils/**/*.js\"",
-    "lint:js": "eslint packages/*/src/ utils/ --max-warnings 0",
+    "lint:js": "eslint packages/*/src/ utils/ .storybook/ --ext js,ts,tsx --max-warnings 0",
     "lint:markdown": "markdownlint API.md DEVELOPMENT.md README.md VERSIONING.md packages/*/src/**/*.md packages/*/src/*.md packages/*/README.md",
     "new": "utils/scripts/create-package.js",
     "prepare": "yarn build",
     "start": "utils/scripts/scoped-npm-command.js --script start",
     "start:demo": "yarn build:demo && live-server demo",
+    "start:storybook": "start-storybook -s ./utils/styleguide/public -p 6006",
```
Comment (jzempel) [inline package.json]: Just to note, both `start:storybook` and `build:storybook` will eventually go away as they become the new `start:demo` and `build:demo`.

File: packages/dropdowns/package.json [hunk 12]
```diff
@@ -14,10 +14,10 @@
     "dist"
   ],
   "scripts": {
-    "build": "../../utils/scripts/build.sh",
-    "build:demo": "../../utils/scripts/build-demo.sh",
-    "start": "../../utils/scripts/start.sh"
+    "build": "../../utils/scripts/build.sh && yarn build:types",
+    "build:types": "tsc -p tsconfig.build.json"
```
Comment (jzempel) [inline packages/dropdowns/package.json]: Reduce scripting footprint by removing `build:types` and just adding to `build`.

File: packages/dropdowns/package.json [hunk 13]
```diff
@@ -14,10 +14,10 @@
     "dist"
   ],
   "scripts": {
-    "build": "../../utils/scripts/build.sh",
-    "build:demo": "../../utils/scripts/build-demo.sh",
-    "start": "../../utils/scripts/start.sh"
```
Comment (jzempel) [inline packages/dropdowns/package.json]: Removing `start` breaks `yarn start` (styleguidist) for dropdowns – not sure this is desired.

File: packages/dropdowns/src/stories/autocomplete.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+              if (Object.prototype.hasOwnProperty.call(changes, 'inputValue')) {
+                setInputValue(changes.inputValue);
+              }
+            }}
+            downshiftProps={{ defaultHighlightedIndex: 0 }}
+          >
+            <Field>
+              <Label>Autocomplete with debounce</Label>
+              <Hint>This example includes basic debounce logic using Hooks</Hint>
+              <Autocomplete>
```
Comment (jzempel) [inline packages/dropdowns/src/stories/autocomplete.stories.tsx]: Use size knob to modify the Autocomplete size in conjunction with the menu.

File: packages/dropdowns/src/stories/menu.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+const GARDEN_PLACEMENT_OPTIONS = Object.values(GARDEN_PLACEMENTS);
+
+storiesOf('Dropdowns.Menu', module)
+  .addDecorator(centered)
+  .add(
+    'default usage',
+    () => (
+      <Dropdown onSelect={action('item-selected')}>
+        <Trigger>
+          <Button>Default Menu</Button>
```
Comment (jzempel) [inline packages/dropdowns/src/stories/menu.stories.tsx]: Use size knob to modify Button in conjunction with Menu.

File: demo/index.html [hunk 16]
```diff
@@ -106,7 +106,7 @@ <h1 class="c-main__title">React Components</h1>
               <a class="u-fg-inherit" href="chrome">Chrome</a>
             </div>
             <div class="u-mb-sm">
-              <a class="u-fg-inherit" href="dropdowns">Dropdowns</a>
+              <a class="u-fg-inherit" href="storybook">Dropdowns</a>
```
Comment (jzempel) [inline demo/index.html]: Make this href navigate directly to the `dropdowns` section. This will also fix the current poor back button behavior. See `react-containers` for details.

**Comment by jzempel** [comment]
Additionally, the breakout of a storybook PR needs Garden designer review.

File: .vscode/settings.json (hunk 4 above)
Comment (austin94) [reply .vscode/settings.json]: I've seem Typescript related config as the exception sometimes due to the miss-match in editor plugins. I'll remove for this iteration.

File: .travis.yml (hunk 3 above)
Comment (austin94) [reply .travis.yml]: D'oh. Moved line 23 to this placement.

File: packages/dropdowns/src/Autocomplete/Autocomplete.tsx (hunk 8 above)
Comment (austin94) [reply packages/dropdowns/src/Autocomplete/Autocomplete.tsx]: This goes beyond the non-null operator though.

`preventDownshiftDefault` is a custom property that Downshift pony-fills to their event handling. Our render-prop wrappings don't show the correct types for this event and `any` is the easiest fix.

File: .eslintignore [hunk 17]
```diff
@@ -0,0 +1 @@
+packages/notifications/index.d.ts
```
[the remark this answers is not in the document]
Comment (austin94) [reply .eslintignore]: Eslint now lints all typescript files. This is an older file that it is fixing to align with our new rules.

File: .lintstagedrc [hunk 18]
```diff
@@ -7,6 +7,20 @@
...
       "git add"
     ],
+    "*.tsx": [
+      "stylelint",
+      "eslint",
+      "jest --config=utils/test/jest.config.js --findRelatedTests",
+      "prettier --write",
+      "git add"
+    ],
+    "*.ts": [
```
[the remark this answers is not in the document]
Comment (austin94) [reply .lintstagedrc]: Huh, tried this originally and didn't find it working. I'll try again.

**Comment by austin94** [comment]
Closing in favor of #371 

I will address all of the relevant comments from here in that PR. Thanks!
